libbeat/processors/add_remote_metadata/add_remote_metadata.go

```diff
@@ -0,0 +1,223 @@
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package add_process_metadata
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/common/atomic"
+	"github.com/elastic/beats/v7/libbeat/processors"
+	conf "github.com/elastic/elastic-agent-libs/config"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/elastic-agent-libs/mapstr"
+)
+
+const name = "add_remote_metadata"
+
+func init() {
+	// We cannot use this as a JS plugin as it is stateful and includes a Close method.
+	processors.RegisterPlugin(name, New)
+}
+
+var (
+	// ErrNoMatch is returned when the event doesn't contain any of the fields
+	// specified in match_pids.
+	ErrNoMatch = errors.New("none of the fields in match_keys found in the event")
+
+	// ErrNoData is returned when metadata for a metadata can't be collected.
+	ErrNoData = errors.New("metadata not found")
+
+	instanceID atomic.Uint32
+)
+
+type addRemoteMetadata struct {
+	config   config
+	mappings mapstr.M
+	src      MetadataGetter
+	cancel   context.CancelFunc
+	log      *logp.Logger
+}
+
+// Resulting processor implements `Close()` to release the cache resources.
+func New(cfg *conf.C) (beat.Processor, error) {
+	config := defaultConfig()
+	err := cfg.Unpack(&config)
+	if err != nil {
+		return nil, fmt.Errorf("fail to unpack the %s configuration: %w", name, err)
+	}
+	new, ok := providers[config.Provider]
+	if !ok {
+		return nil, fmt.Errorf("unknown provider: %s", config.Provider)
+	}
+	src, cancel, err := new(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("fail to unpack the %s configuration for provider %s: %w", name, config.Provider, err)
+	}
+
+	// Logging (each processor instance has a unique ID).
+	id := int(instanceID.Inc())
+	log := logp.NewLogger(name).With("instance_id", id)
+
+	mappings, err := config.getMappings()
+	if err != nil {
+		return nil, fmt.Errorf("error unpacking %v.target_fields: %w", name, err)
+	}
+
+	p := addRemoteMetadata{
+		config:   config,
+		src:      src,
+		cancel:   cancel,
+		mappings: mappings,
+		log:      log,
+	}
+	return &p, nil
+}
+
+// MetadatGetter is the interface implemented by metadata providers.
+type MetadataGetter interface {
+	GetMetadata(key string) (interface{}, error)
+}
+
+var providers = map[string]func(*conf.C) (MetadataGetter, context.CancelFunc, error){
+	"map": newMapProvider,
+}
+
+// mapProvider is a simple providers based on a static map look-up.
```

Comment:
Not sure I understand it correctly but the comment didn't make any sense.
```suggestion
// mapProvider is a set of providers based on a static map look-up.
```

Reply:
s/providers/provider/